### PR TITLE
control-service: extend GraphQL execution data fetcher

### PR DESCRIPTION
**Pull request by mivanov1988**
Extended the existing GraphQL data fetcher for executions
(com.vmware.taurus.service.graphql.ExecutionDataFetcher)
to support filtering and sorting. The filtering and sorting
work at the database level for better performance
and resource (memory, network traffic, etc.) utilization.
The queries are be built dynamically based on the filters.

Sample GraphQL query:
```
{{ baseUrl  }}/data-jobs/for-team/no-team-specified/jobs?query={
  executions(pageNumber: 1, pageSize: 100, filter: {statusIn: [RUNNING, SUBMITTED]}, order: {property: "status", direction: DESC}]) {
    content {
          id
          startTime
          endTime
          status
          jobName
    }
  }
}
```
Sample GraphQL response:
```json
{
  "errors": [],
  "data": {
    "content": [
     {
        "id": "tms-staging-vdk-1633680894-1634638440",
        "startTime": "2021-10-19T10:14:47Z",
        "endTime": null,
        "status": "running",
        "jobName": "test_job1",
      },
     {
        "id": "tms-staging-vdk-1643242444-2131231312",
        "startTime": "2021-10-19T10:14:47Z",
        "endTime": null,
        "status": "submitted",
        "jobName": "test_job3",
      },
     {
        "id": "tms-staging-vdk-121321314-12312312333",
        "startTime": "2021-10-19T10:14:47Z",
        "endTime": null,
        "status": "running",
        "jobName": "test_job2",
      }
    ],
    "totalItems": 0,
    "totalPages": 0
  }
}
```

Testing: unit tests

Signed-off-by: Miroslav Ivanov miroslavi@vmware.com

### CODE REVIEW

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/JobExecutionRepository.java [hunk 1]
```diff
@@ -27,7 +30,10 @@
  * <p>
  * JobExecutionRepositoryIT validates some aspects of the behavior
  */
-public interface JobExecutionRepository extends JpaRepository<DataJobExecution, String> {
+public interface JobExecutionRepository extends
+      JpaRepository<DataJobExecution, String>,
+      JpaSpecificationExecutor<DataJobExecution>,
```
Comment (mrMoZ1) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/JobExecutionRepository.java:35]: JpaRepository already extends PagingAndSorting:
https://docs.spring.io/spring-data/jpa/docs/current/api/org/springframework/data/jpa/repository/JpaRepository.html

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/datajobs/ToApiModelConverter.java [hunk 2]
```diff
@@ -124,7 +125,12 @@ public static DataJobQueryResponse toDataJobPage(ExecutionResult executionResult
          return dataJobQueryResponse;
       }
 
-      LinkedHashMap<String, Object> nestedExecutionResultData = executionResultData.get("jobs");
+      String executionResultDataKey = GraphQLUtils.QUERIES
+            .stream()
+            .filter(query -> executionResultData.containsKey(query))
+            .findFirst()
+            .get();
```
Comment (mrMoZ1) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/datajobs/ToApiModelConverter.java:132]: Is this safe from NoSuchElementException?

Comment (mrMoZ1) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/datajobs/ToApiModelConverter.java:132]:
```suggestion
            .orElse(null);;
```

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/graphql/model/DataJobExecutionOrder.java [hunk 3]
```diff
@@ -5,13 +5,40 @@
 
 package com.vmware.taurus.service.graphql.model;
 
-import lombok.AllArgsConstructor;
+import com.google.common.collect.ImmutableSet;
+import lombok.Builder;
 import lombok.Data;
 import org.springframework.data.domain.Sort;
 
+import com.vmware.taurus.service.model.DataJobExecution_;
```
[the remark this answers is not in the document]
Comment (mivanov1988) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/graphql/model/DataJobExecutionOrder.java:13]: It comes from Hibernate tool called [jpamodelgen](https://docs.jboss.org/hibernate/stable/jpamodelgen/reference/en-US/html_single/) (org.hibernate:hibernate-jpamodelgen). It is a metadata model generator.

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/graphql/ExecutionDataFetcher.java [hunk 4]
```diff
@@ -33,13 +27,41 @@
...
 /**
  * Data fetcher class for Data Job Executions
- *
+ * <p>
  * Data fetchers are classes that provides to graphql api the needed data while it modify the source data:
  * By providing list of data jobs it alter each job and attach its execution while reading requested
  * information from graphql query to specify how many executions, are they sorted by specific field, etc.
- *
+ * <p>
  * Currently, execution data fetcher does not provide filtering by specifying fields due
```
[the remark this answers is not in the document]
Comment (mivanov1988) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/graphql/ExecutionDataFetcher.java:65]: It refers to populateExecutions(). I moved the comment there.

**Review by mrMoZ1 — COMMENTED** [review]
Overall LGTM, but someone else should approve.

**Comment by tozka** [comment]
```
{{ baseUrl  }}/data-jobs/for-team/no-team-specified/jobs?query={
  executions(pageNumber: 1, pageSize: 100, filter: {statusIn: [RUNNING, SUBMITTED]}, order: {property: "status", direction: DESC}]) {
    content {
          id
          startTime
          endTime
          status
          jobName
    }
  }
}
```
The sample SQL query is different from the API doc ???? And it does not look like valid JSON (according to https://jsonlint.com/). So is it correct?

```
filter: [{
             property: "deployments.executions.startTime",
             sort: "DESC",
           }],
```

**Comment by mivanov1988** [comment]
> ```
> {{ baseUrl  }}/data-jobs/for-team/no-team-specified/jobs?query={
>   executions(pageNumber: 1, pageSize: 100, filter: {statusIn: [RUNNING, SUBMITTED]}, order: {property: "status", direction: DESC}]) {
>     content {
>           id
>           startTime
>           endTime
>           status
>           jobName
>     }
>   }
> }
> ```
> 
> The sample SQL query is different from the API doc ???? And it does not look like valid JSON (there's missing [ bracket) . So is it correct?
> 
> Per API doc is:
> 
> ```
> filter: [{
>              property: "deployments.executions.startTime",
>              sort: "DESC",
>            }],
> ```

The documentation refers to `jobs` query. The `executions` query filters are different, so the sample is correct.

**Comment by tozka** [comment]
> The documentation refers to `jobs` query. The `executions` query filters are different, so the sample is correct.

Sorry. I copied only a snippet so perhaps it was not clear. Here si the full [doc](http://ac6502b637b6641b9bbb6ca95c2df190-1965366291.us-west-1.elb.amazonaws.com:8092/swagger-ui.html#/Data%20Jobs/jobsQuery)

```

Data jobs execution could also be searched by providing arguments to the execution field. Same as parent query arguments, the pageNumber and pageSize arguments are required! Page number should be a number greater than 1, and pageSize should be between 1 and 100 results (per page). You can also filter using the similar object structure as the parent query, but currently filtering is not supported, you can only provide field for sorting. This query will search
 {
   jobs(
     pageNumber: 1,
     pageSize: 25,
   ) {
     content {
       jobName,
       deployments {
         id
         executions(
           pageNumber: 1,
           pageSize: 5
           filter: [{
             property: "deployments.executions.startTime",
             sort: "DESC",
           }],
         ) {
           id           status           startTime           endTime         {
       }
     }
   }
 }

```

But in any case - on whatever level it is I think the filtering and ordering syntax should be the same

**Comment by mivanov1988** [comment]
> > The documentation refers to `jobs` query. The `executions` query filters are different, so the sample is correct.
> 
> Sorry. I copied only a snippet so perhaps it was not clear. Here si the full [doc](http://ac6502b637b6641b9bbb6ca95c2df190-1965366291.us-west-1.elb.amazonaws.com:8092/swagger-ui.html#/Data%20Jobs/jobsQuery)
> 
> ```
> 
> Data jobs execution could also be searched by providing arguments to the execution field. Same as parent query arguments, the pageNumber and pageSize arguments are required! Page number should be a number greater than 1, and pageSize should be between 1 and 100 results (per page). You can also filter using the similar object structure as the parent query, but currently filtering is not supported, you can only provide field for sorting. This query will search
>  {
>    jobs(
>      pageNumber: 1,
>      pageSize: 25,
>    ) {
>      content {
>        jobName,
>        deployments {
>          id
>          executions(
>            pageNumber: 1,
>            pageSize: 5
>            filter: [{
>              property: "deployments.executions.startTime",
>              sort: "DESC",
>            }],
>          ) {
>            id           status           startTime           endTime         {
>        }
>      }
>    }
>  }
> ```
> 
> But in any case - on whatever level it is I think the filtering and ordering syntax should be the same

If we want to keep the syntax we

**Comment by mivanov1988** [comment]
> > The documentation refers to `jobs` query. The `executions` query filters are different, so the sample is correct.
> 
> Sorry. I copied only a snippet so perhaps it was not clear. Here si the full [doc](http://ac6502b637b6641b9bbb6ca95c2df190-1965366291.us-west-1.elb.amazonaws.com:8092/swagger-ui.html#/Data%20Jobs/jobsQuery)
> 
> ```
> 
> Data jobs execution could also be searched by providing arguments to the execution field. Same as parent query arguments, the pageNumber and pageSize arguments are required! Page number should be a number greater than 1, and pageSize should be between 1 and 100 results (per page). You can also filter using the similar object structure as the parent query, but currently filtering is not supported, you can only provide field for sorting. This query will search
>  {
>    jobs(
>      pageNumber: 1,
>      pageSize: 25,
>    ) {
>      content {
>        jobName,
>        deployments {
>          id
>          executions(
>            pageNumber: 1,
>            pageSize: 5
>            filter: [{
>              property: "deployments.executions.startTime",
>              sort: "DESC",
>            }],
>          ) {
>            id           status           startTime           endTime         {
>        }
>      }
>    }
>  }
> ```
> 
> But in any case - on whatever level it is I think the filtering and ordering syntax should be the same

In order to keep the syntax we should make the following changes:
1. Refactor the schema: 
```
schema {
    query: Query
}

type Query {
    jobs(pageNumber: Int = 0, pageSize: Int = 20, filter: [Predicate], search: String): DataJobPage
    executions(pageNumber: Int, pageSize: Int, filter: [Predicate]): DataJobExecutionResponse
}

input Predicate {
    property: String,
    operation: Operation = EQ,
    pattern: [String!],
    sort: Direction
}

enum Direction {
    ASC,
    DESC
}

enum Operation {
    EQ,
    GT,
    GTE,
    IN
}
```

2. Refactor the JobExecutionFilterSpec to support dynamic filters:
```java
public class JobExecutionFilterSpec implements Specification<DataJobExecution> {

   Optional<List<Filter>> filters;

   Map<String, SingularAttribute> supportedFields = Map.of(
         DataJobExecution_.status.getName(), DataJobExecution_.status,
         DataJobExecution_.startTime.getName(), DataJobExecution_.startTime,
         DataJobExecution_.endTime.getName(), DataJobExecution_.endTime,
         ...
   );

   @Override
   public Predicate toPredicate(Root<DataJobExecution> root, CriteriaQuery<?> query, CriteriaBuilder builder) {
      List<Predicate> predicates = new ArrayList<>();

      if (filters.isPresent()) {
         List<Filter> filters = this.filters.get();

         for (Filter filter : filters) {
            if (supportedFields.containsKey(filter.getProperty())) {
               SingularAttribute field = supportedFields.get(filter.getProperty());

               switch (filter.getOperation()) {
                  case EQ:
                     predicates.add(builder.equal(root.get(field.getName()), filter.getPattern().get(0)));
                  case GT:
                     predicates.add(builder.greaterThan(root.get(field.getName()), filter.getPattern().get(0)));
                  case GTE:
                     predicates.add(builder.greaterThanOrEqualTo(root.get(field.getName()), filter.getPattern().get(0)));
                  case IN: {
                     predicates.add(root.get(field.getName()).in(filter.getPattern()));
                  }
               }
            }
         }
}
```
3. Since the `pattern` is going to be List\<String\> we have to introduce converters and validators for each field like: 
```
status -> StatusEnumConverter, 
startTime -> OffsetDateTimeConverter,
...
```
4. Since the `pattern` is going to be changed from String to List\<String\> we have to refactor all existing GraphQL code that uses it.

It is possible to keep the syntax but it will took me at least 2-3 days more.

**Comment by tozka** [comment]
> It is possible to keep the syntax but it will take me at least 2-3 days more. Also this schema has already been approved in #478.

I've missed that in #478 . We can live with the inconsistencies. But I think our UX would be better if we have a consistent way to specify filtering and sorting for resources. Consider if we should open a ticket for that.

**Comment by tozka** [comment]
Should we add integration test ? I deployed the code from the branch locally and the sample query failed with an exception for me?

**Comment by mivanov1988** [comment]
> Should we add integration test ? I deployed the code from the branch locally and the sample query failed with an exception for me?

It is going to be added as part of #511. Could you please send me the error message? Also you could try with the following query:

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/graphql/model/GraphQLResponse.java [hunk 5]
```diff
@@ -3,14 +3,18 @@
...
+import lombok.Builder;
 import lombok.Data;
 
 import java.util.List;
 
 @Data
-public class DataJobPage {
+@Builder
+@AllArgsConstructor
+public class GraphQLResponse {
```
Comment (tozka) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/graphql/model/GraphQLResponse.java:17]: Why rename it? It still returns Data Jobs - only information about their executions.

**Review by tozka — APPROVED** [review]
Looks good to me. Thanks for doing the change.

**Comment by tozka** [comment]
> > Should we add integration test ? I deployed the code from the branch locally and the sample query failed with an exception for me?
> 
> It is going to be added as part of #511. Could you please send me the error message? Also you could try with the following query:
> 
> ```
> {
>   executions(pageSize: 1, pageNumber: 1, filter: {startTimeGte: "1996-12-19T16:39:57-08:00", statusIn: [RUNNING, FAILED]}) {
>     content {
>       id
>       startTime
>     }
>     totalPages
>     totalItems
>   }
> }
> ```

Yes, this works now.

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/graphql/model/GraphQLResponse.java (hunk 5 above)
Comment (mivanov1988) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/graphql/model/GraphQLResponse.java:17]: Yes, you have the point.

**Comment by mivanov1988** [comment]
> > > Should we add integration test ? I deployed the code from the branch locally and the sample query failed with an exception for me?
> > 
> > 
> > It is going to be added as part of #511. Could you please send me the error message? Also you could try with the following query:
> > ```
> > {
> >   executions(pageSize: 1, pageNumber: 1, filter: {startTimeGte: "1996-12-19T16:39:57-08:00", statusIn: [RUNNING, FAILED]}) {
> >     content {
> >       id
> >       startTime
> >     }
> >     totalPages
> >     totalItems
> >   }
> > }
> > ```
> 
> Yes, this works now.

Sorry, the example in the description has

**Pull request merged**